chain/chain/src/chain.rs

```diff
@@ -788,74 +791,346 @@ impl Chain {
+            let block_header = self.get_block_header(&hash)?;
+            if block_header.inner.height <= height_included {
+                receipt_proof_response.push(incoming_receipt_proof.clone());
+            }
+        }
+        let receipts = self.runtime_adapter.collect_receipts_from_response(&receipt_proof_response);
+
+        let mut transactions = block_transactions;
+        transactions.extend(chunk.transactions.iter().cloned());
+
+        let gas_limit = chunk.header.inner.gas_limit;
+        let mut apply_result = self
+            .runtime_adapter
+            .apply_transactions(
+                shard_id,
+                &chunk.header.inner.prev_state_root,
+                chunk.header.height_included,
+                &chunk.header.inner.prev_block_hash,
+                &block_header.hash,
+                &receipts,
+                &transactions,
+                block_header.inner.gas_price,
+            )
+            .map_err(|e| ErrorKind::Other(e.to_string()))?;
+
+        // Saving the state.
         let mut chain_store_update = self.store.store_update();
-        chain_store_update.save_chunk_extra(&prev_block_hash, shard_id, prev_extra);
-        chain_store_update.save_outgoing_receipt(
-            &outgoing_receipts.0,
-            shard_id,
-            outgoing_receipts.1,
+        chain_store_update.save_trie_changes(apply_result.trie_changes);
+        let chunk_extra = ChunkExtra::new(
+            &apply_result.new_root,
+            apply_result.validator_proposals,
+            apply_result.total_gas_burnt,
+            gas_limit,
         );
-        for incoming_receipt in incoming_receipts {
+        chain_store_update.save_chunk_extra(&block_header.hash, shard_id, chunk_extra);
+        // Saving outgoing receipts.
+        let mut outgoing_receipts = vec![];
+        for (_receipt_shard_id, receipts) in apply_result.receipt_result.drain() {
+            outgoing_receipts.extend(receipts);
+        }
+        chain_store_update.save_outgoing_receipt(&block_header.hash(), shard_id, outgoing_receipts);
+        // Saving transaction results.
+        for tx_result in apply_result.transaction_results.into_iter() {
+            chain_store_update.save_transaction_result(&tx_result.hash, tx_result.result);
+        }
+        // Saving all incoming receipts.
+        for receipt_proof_response in incoming_receipts_proofs {
             chain_store_update.save_incoming_receipt(
-                &incoming_receipt.0,
+                &receipt_proof_response.0,
                 shard_id,
-                incoming_receipt.1,
+                receipt_proof_response.1,
             );
         }
+        // Committing all the state.
         chain_store_update.commit()?;
 
+        let mut current_height = chunk.header.height_included;
+        loop {
+            current_height += 1;
+            let block_header = self.get_header_by_height(current_height)?.clone();
```

Comment:
will fail if the height is skipped. check if height is present and move to the next iter if not